planner/core/rule_eliminate_projection.go

```diff
@@ -109,6 +118,24 @@ func eliminatePhysicalProjection(p PhysicalPlan) PhysicalPlan {
 type projectionEliminator struct {
 }
 
+// ReplaceColumnOfExprInPhysicalProjection replaces column of expression by another PhysicalProjection.
+func ReplaceColumnOfExprInPhysicalProjection(expr expression.Expression, proj *PhysicalProjection, schema *expression.Schema) expression.Expression {
+	switch v := expr.(type) {
+	case *expression.Column:
+		idx := schema.ColumnIndex(v)
+		if idx != -1 && idx < len(proj.Exprs) {
+			return proj.Exprs[idx]
+		}
+	case *expression.CorrelatedColumn:
+		ReplaceColumnOfExprInPhysicalProjection(&v.Column, proj, schema)
+	case *expression.ScalarFunction:
+		for i := range v.GetArgs() {
```

Comment:
Though it's the last phase of optimizing so directly replace the args of one expression should be considered as a safe operation. But we still suggest that create a new expression instead.

Reply:
> Though it's the last phase of optimizing so directly replace the args of one expression should be considered as a safe operation. But we still suggest that create a new expression instead.

@winoros PTAL.